### ISSUE TITLE
FIM with >5 regions?

### ISSUE THREAD

**Issue by EvanKomp** [Bitbol-Lab/ProtMamba-ssm#19]
Hello, thanks for your work.

I am wondering if FIM with >5 patches is possible without retraining? Currently the vocabulary does not include more them `<mask-5>`. Would cheating this by having multiple patches have the same token screw everything up?

I think your tool would be **extremely useful** if so. Often we know what parts of eg. an enzyme we need to keep stationary to maintain a certain function - those positions likely result in many more than 5 discontinuous sections. It would be great to be able to fill in the rest of the protein around those fixed residues based on specific context sequences that have the properties we care about.

Thanks for any insight.

**Comment by damiano-sg** [comment]
Hello @EvanKomp sorry for the late reply, unfortunately the model was trained with hardcoded mask tokens up until mask-5, so you'll have to retrain/finetune it if you want to use more masks. As you correctly say, this doesn't mean that you can mask only up to 5 residues, it just means that you can only mask up to 5 non-contiguous regions in the sequence.

An alternative solution to your problem, in case you want to mask more than 5 non-contiguous regions and condition on homolog sequences, is to use conditional diffusion pLMs like RAG-ESM that we present in this paper: https://journals.aps.org/prxlife/abstract/10.1103/db1b-hy16

You can find all the code in this repo: https://github.com/Bitbol-Lab/rag-esm.git

**Comment by EvanKomp** [comment]
@damiano-sg Copy that - however I believe that having multiple context sequences is an important feature of this tool as compared to your new one. 

I am wondering if, for say N context sequences, one could average the results of all N? The most obvious / dumbest way to do this would be summing the logits or something. Seems like it might be possible to (probably quite slow ...)

```
for layer in model:
    for c in context sequences:
        attn_weights, values = attn(c, query)
    compute and normalize messages over N context sequences
    skip connection, norm, FFNN
```
Output logits are now computed using N context sequences without training a different architecture? Maybe dumb but from an enzyme engineering persepective, 1) control over specific residues while 2) still having evolutionary context (many sequences) is key.

**Comment by damiano-sg** [comment]
@EvanKomp that makes sense to me, I agree that it's going to be very slow though, I think it's worth trying anyways. Maybe finetuning the model on the task could also give some better results, because I fear that it might be a little too "out of distribution" otherwise.
